File: modules/commodetto/commodettoParseBMF.js
/*
 * Copyright (c) 2016-2017  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK Runtime.
 * 
 *   The Moddable SDK Runtime is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 * 
 *   The Moddable SDK Runtime is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 * 
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with the Moddable SDK Runtime.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


/*
	Commodetto BMF loader

		parses BMF data structure
		puts needed header fields and glyph metrics into ArrayBuffer for use by native rendering code

		http://www.angelcode.com/products/bmfont/doc/file_format.html


		N.B. version 4 is a Moddable variation which stores 4-bit compressed gray bitmaps
			in the file rather than referencing an externa; bitmap. x & y fields are replaced with a 32-bit
			little-endian offset to the compressed glyph (offset is from start of character code entry,
			not the start of the file)

*/

function parse() @ "xs_parseBMF"

export default parse;

/*
export default function parseBMF(font)
{
	let parser = new Parser(font);
	return parser.parse();
}

class Parser {
	constructor(buffer) {
		this.bytes = new Uint8Array(buffer);
	}

	seekTo(position) {
		this.position = position;
	}
	seekBy(delta) {
		this.position += delta;
	}
	readU32() {
		let value = this.bytes[this.position] |  (this.bytes[this.position + 1] << 8) | (this.bytes[this.position + 2] << 8) | (this.bytes[this.position + 3] << 8);
		this.position += 4;
		return value;
	}
	readS16() {
		let value = (this.bytes[this.position]) | (this.bytes[this.position + 1] << 8);
		if (value > 32767) value -= 65536;
		this.position += 2;
		return value;
	}
	readU8() {
		return this.bytes[this.position++];
	}

	parse() {
		let bytes = this.bytes, buffer = bytes.buffer;

		if ((0x42 != bytes[0]) || (0x4D != bytes[1]) || (0x46 != bytes[2]) || ((3 != bytes[3]) && (4 != bytes[3])))
			throw new Error("Invalid BMF header");

		this.seekTo(4);

		// skip block 1
		if (1 != this.readU8())
			throw new Error("can't find info block");

		this.seekBy(this.readU32());

		// get lineHeight from block 2
		if (2 != this.readU8())
			throw new Error("can't find common block");
		let size = this.readU32();

		buffer.height = this.readS16();
		buffer.ascent = this.readS16();
		this.seekBy(size - 4);

		// skip block 3
		if (3 != this.readU8())
			throw new Error("can't find pages block");

		this.seekBy(this.readU32());

		// use block 4
		if (4 != this.readU8())
			throw new Error("can't find chars block");

		buffer.position = this.position;		// position of size of chars table

		size = this.readU32();
		if (size % 20)
			throw new Error("bad chars block size");
//@@ this scan of the character table should be unnecessary on device... font data has already checked at buildtime
		let firstChar = this.readU32();
		let lastChar = firstChar + (size / 20);
		this.seekBy(20 - 4);
		for (let i = firstChar + 1; i < lastChar; i++) {
			let c = this.readU32();
			if (c != i)
				throw new Error("gap detected - chars must be consecutive");
			this.seekBy(20 - 4);
		}

		buffer.firstChar = firstChar;
		buffer.lastChar = lastChar;

		return buffer;
	}
};

Object.freeze(Parser.prototype);

*/

File: modules/commodetto/commodettoParseBMF.c
/*
 * Copyright (c) 2016-2021  Moddable Tech, Inc.
 *
 *   This file is part of the Moddable SDK Runtime.
 * 
 *   The Moddable SDK Runtime is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU Lesser General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 * 
 *   The Moddable SDK Runtime is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 * 
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with the Moddable SDK Runtime.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


#include "xsPlatform.h"

#include "xsmc.h"
#include "mc.xs.h"			// for xsID_ values

//@@ can read beyond end of provided buffer on malformed data
void xs_parseBMF(xsMachine *the)
{
	unsigned char *bytes, *start;
	uint32_t size;
	xsUnsignedValue byteLength;
	int charCount;

	xsmcGetBufferReadable(xsArg(0), (void **)&bytes, &byteLength);
	start = bytes;

	if ((0x42 != c_read8(bytes + 0)) || (0x4D != c_read8(bytes + 1)) || (0x46 != c_read8(bytes + 2)) || ((3 != c_read8(bytes + 3)) && (4 != c_read8(bytes + 3))))
		xsUnknownError("Invalid BMF header");
	bytes += 4;

	// skip block 1
	if (1 != c_read8(bytes))
		xsUnknownError("can't find info block");
	bytes += 1;

	bytes += 4 + c_read32(bytes);

	// get lineHeight from block 2
	if (2 != c_read8(bytes))
		xsUnknownError("can't find common block");
	bytes += 1;

	size = c_read32(bytes);
	bytes += 4;

	xsmcSetInteger(xsResult, c_read16(bytes));
	bytes += 2;
	xsmcDefine(xsArg(0), xsID_height, xsResult, xsDontDelete | xsDontSet);

	xsmcSetInteger(xsResult, c_read16(bytes));
	bytes += 2;
	xsmcDefine(xsArg(0), xsID_ascent, xsResult, xsDontDelete | xsDontSet);

	bytes += 2 + 2;		// scaleW and scaleH
	if (1 != c_read16(bytes))	// pages
		xsUnknownError("not single page");

	bytes += size - 8;

	// skip block 3
	if (3 != c_read8(bytes))
		xsUnknownError("can't find pages block");
	bytes += 1;

	bytes += 4 + c_read32(bytes);

	// use block 4
	if (4 != c_read8(bytes))
		xsUnknownError("can't find chars block");
	bytes += 1;

	xsmcSetInteger(xsResult, bytes - start);
	xsmcDefine(xsArg(0), xsID_position, xsResult, xsDontDelete | xsDontSet);

	size = c_read32(bytes);
	if (size % 20)
		xsUnknownError("bad chars block size");
	charCount = size / 20;

	xsmcSetInteger(xsResult, charCount);
	xsmcDefine(xsArg(0), xsID_charCount, xsResult, xsDontDelete | xsDontSet);

	xsResult = xsArg(0);
}
